platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java

```diff
@@ -102,6 +103,19 @@ protected Configuration doCreate(String name) {
         return defaultConfigurationFactory.create(name, this, resolutionStrategyFactory, rootComponentMetadataBuilder, ConfigurationRoles.LEGACY);
     }
 
+    @Override
+    protected NamedDomainObjectProvider<Configuration> createDomainObjectProvider(String name, @Nullable Action<? super Configuration> configurationAction) {
+        // Called by `register` for registering legacy configurations.
+        // We override to set the public type to `DefaultUnlockedConfiguration`, allowing us to filter for unlocked configurations with `withType`
+
+        assertElementNotPresent(name);
+        NamedDomainObjectProvider<Configuration> provider = Cast.uncheckedCast(
+            getInstantiator().newInstance(AbstractNamedDomainObjectContainer.NamedDomainObjectCreatingProvider.class, DefaultConfigurationContainer.this, name, DefaultUnlockedConfiguration.class, configurationAction)
```

Comment:
Will the `DefaultUnlockedConfiguration` type be visible to users here?

Should we create a public `UnlockedConfiguration` interface in that case?